modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  The :mod:`Initialization` module provides functions which enable users to
+  move-initialize and deinitialize values.
+
+  The functionality provided by this module can be used to implement
+  collections in a manner similar to those implemented by the Chapel standard
+  modules (such as :mod:`List` or :mod:`Map`).
+*/
+module Initialization {
+
+  private inline proc _move(ref dst: ?t, const ref src: t) {
+    __primitive("=", dst, src);
+  }
+
+  private inline proc _destroy(ref r) {
+    chpl__autoDestroy(r);
+  }
+
+  private inline proc _needsAutoDestroy(type t) param: bool {
+    return __primitive("needs auto destroy", t);
+  }
+
+  /*
+    Check to see if a given type supports deinitialization.
```

Comment:
> if a given type has a non-empty deinitalizer

That leaves a question about `proc MyRecord.deinit() { }` because the compiler adds calls to field deinits in that case. How about "Check to see if a given type needs to be deinitialized".